src/executor/cursor.rs

```diff
@@ -101,27 +101,55 @@ where
                         .add(f(c2, v2)),
                 )
                 .add(f(c1, v1)),
-            (Identity::Many(col_vec), ValueTuple::Many(value_vec))
-                if col_vec.len() == value_vec.len() =>
+            (Identity::Many(col_vec), ValueTuple::Many(val_vec))
+                if col_vec.len() == val_vec.len() =>
             {
-                let len = col_vec.len();
-                let mut cond_any = Condition::any();
-                for n in (1..=len).rev() {
-                    let mut cond_all = Condition::all();
-                    for (i, (col, value)) in
-                        col_vec.iter().zip(value_vec.iter()).enumerate().take(n)
-                    {
-                        let v = value.clone();
-                        let expr = if i != (n - 1) {
-                            Expr::col((SeaRc::clone(&self.table), SeaRc::clone(col))).eq(v)
-                        } else {
-                            f(col, v)
-                        };
-                        cond_all = cond_all.add(expr);
-                    }
-                    cond_any = cond_any.add(cond_all);
-                }
-                cond_any
+                // The length of `col_vec` and `val_vec` should be equal and is denoted by "n".
+                //
+                // The elements of `col_vec` and `val_vec` are denoted by:
+                //   - `col_vec`: "col_1", "col_2", ..., "col_n-1", "col_n"
+                //   - `val_vec`: "val_1", "val_2", ..., "val_n-1", "val_n"
+                //
+                // The general form of the where condition should have "n" number of inner-AND-condition chained by an outer-OR-condition.
+                // The "n"-th inner-AND-condition should have exactly "n" number of column value expressions,
+                // to construct the expression we take the first "n" number of column and value from the respected vector.
+                //   - if it's not the last element, then we construct a "col_1 = val_1" equal expression
+                //   - otherwise, for the last element, we should construct a "col_n > val_n" greater than or "col_n < val_n" less than expression.
+                // i.e.
+                // WHERE
+                //   (col_1 = val_1 AND col_2 = val_2 AND ... AND col_n > val_n)
+                //   OR (col_1 = val_1 AND col_2 = val_2 AND ... AND col_n-1 > val_n-1)
+                //   OR (col_1 = val_1 AND col_2 = val_2 AND ... AND col_n-2 > val_n-2)
+                //   OR ...
+                //   OR (col_1 = val_1 AND col_2 > val_2)
+                //   OR (col_1 > val_1)
```

Comment:
Greatly appreciated!

Reply:
A comprehensive testing on its way :P

Reply:
Check this. The algorithm in action. https://github.com/SeaQL/sea-orm/pull/1508/commits/2bec23a37c6182588869edf51ae3210df9dde502